---
title: "Plagues of the Past and Present"
authors: "Katherine Eaton"
dataset: "https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=map&legend=closed"
abstract: "
'The Plague' is a mystifying and devastating disease that has re-emerged multiple times throughout history. From the Plague of Justinian (6th century), through the centuries long Black Death (14th-19th century), to the Madagascar Plague Outbreak (2017), this infectious disease has resulted in exceptional mortality and societal upheaval.

### While our knowledge of the plague has grown tremendously, there are many outstanding questions:
<br><br>
1. <b>Where</b> did the ancient pandemics originate?
<br>
2. <b>When</b> might dormant plague zones re-activate?
<br>
3. <b>Why</b> has the plague gone extinct in some regions?
<br>
4. <b>How</b> does a community respond to losing a large portion of its population?
<br><br>

### The mysteries of the plague bring together researchers from a wide variety of disciplines from art history to microbiology. Each field contributes a unique perspective to the questions of how people are affected and cope with this disease, as well as where, when, and why it suddenly appears...

## Table of Contents

###
<pre>
   [Origins](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=1)
<br>
   [Evolution](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=4)
<br>
   [Ecology](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=5)
<br>
   [Digital Scholarship](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=7)
</pre>
</p>
"
---


<!----------------------------------------------------------------------------->
<!--                     1 : LS : Origins Ancient Plague                     -->
<!----------------------------------------------------------------------------->
# [Contentious Origins](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=map)
#### The plagues of the ancient world have intriguing geographic origins.

#### Early hypotheses on the 'origins' of the ancient plagues varied greatly including regions in [Central and East Africa](https://doi.org/10.2307/600071)[<sup>1</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8), [the Middle East, as well as Western and Central Asia](https://boydellandbrewer.com/the-black-death-1346-1353-the-complete-history.html)[<sup>2</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8).<br>

#### The historical sources used to [estimate plague mortality](https://www.springer.com/gp/book/9783540257943)[<sup>3</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8) consist of documents such as parish records, financial transactions, wills, and local testimonies. These documents are used to trace the progressive movement of plague between populations and are visualized in the form of spread maps.

#### However, these source documents can be of poor quality or are sparsely found leading to [high levels of uncertainty](https://doi.org/10.1111/j.1467-9671.2012.01369.x)[<sup>4</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8). As a result, the historical sources used in the construction of these maps are experiencing [renewed critical re-examination](https://www.routledge.com/Spatial-Analysis-in-Health-Geography/Kanaroglou-Delmelle/p/book/9781138546615)[<sup>5</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8).

<!----------------------------------------------------------------------------->
<!--                     1 : RS : Origins Maps                               -->
<!----------------------------------------------------------------------------->
```auspiceMainDisplayMarkdown

<div>
  <a href="https://doi.org/10.13140/RG.2.2.25362.25281">
    <img alt="microscopy image of coronaviruses" width="500" src="https://www.researchgate.net/profile/Anne_Laudisoit2/publication/315837122/figure/fig1/AS:652961457897473@1532689551350/a-leftThe-origin-and-spread-of-Justinian-plague.png"/>
      <pre>                         Spread of the Justinian Plague [6]                         </pre>
  </a>

  <a href="https://boydellandbrewer.com/the-black-death-1346-1353-the-complete-history.html">
    <img alt="illustration of a coronavirus" width="600" src="https://www.historytoday.com/sites/default/files/blackdeathmap.jpg"/>
      <pre>                                  Spread of the Black Death Plague [2]                                </pre>
  </a>

</div>

```


<!----------------------------------------------------------------------------->
<!--                     2 : LS : Origins New Lens                           -->
<!----------------------------------------------------------------------------->
# [New Perspectives](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=map&animate=1400-01-01,2017-01-01,0,1,30000)
#### In contrast to these initial historical perspectives, genetic evidence offers a different narrative.

#### Advances in DNA sequencing have made it possible to rapidly sequence the genome of the plague bacterium, *Yersinia pestis*, from outbreak sources across the globe.

#### DNA is unique in that it provides a continual record of change over time. That includes both mutations occurring within an organism's lifetime as well as those that appeared in an ancestral population. Because of this unique characteristic, DNA can be used to make inferences about events occurring in the very distant past. One example presented here is linking ancient evolutionary events to geographic locations.

#### The accompanying spread map visualizes 151 DNA samples of modern *Y. pestis* (1947-2016). The underlying analyses calculate relatedness between plague bacteria and track the mutations that have occurred over time and space. Estimates are then provided on the possible locations of plague in the distant past, suggesting the most likely regions to be [in or near East Asia]((https://doi.org/10.1038/ng.705) [<sup>7</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8).

#### [Restart Spread Animation](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=2).

#### However, modern DNA as a form of evidence, similar to the historical sources discussed previously, suffers from several key flaws. Notably, the strong geographic bias towards where <i>modern</i> plague epidemics occur has greatly restricted our view of the past.

<!----------------------------------------------------------------------------->
<!--                     3 : LS : Origins Ancient DNA                        -->
<!----------------------------------------------------------------------------->
# [Ancient DNA](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=map&country=sweden)
#### And so the picture grows murkier still. Ancient plague DNA retrieved from victims of historic epidemics provides new windows into the past.
<a href="https://www.history.com/news/is-the-black-death-the-ancestor-of-all-modern-plagues">
  <img src="https://www.history.com/.image/c_limit%2Ccs_srgb%2Cq_auto:good%2Cw_686/MTU3ODc4NTk4NjgzOTI4Mjg3/image-placeholder-title.webp" width="70%">
        <p style='text-align: center;'>London's East Smithfield "plague pits", Black Death.</p>
</a>

#### A recent study identified the plague bacterium in skeletal remains from [Sweden 4,900 years ago](https://doi.org/10.1016/j.cell.2018.11.005)[<sup>8</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8). Given that Europe was largely excluded from being a candidate origin of plague, this finding is particularly surprising. New hypotheses are constantly emerging concerning the spread of plague across Eurasia. The case is anything but closed.
<a href="https://doi.org/10.1016/j.cell.2018.11.005">
  <img src="https://raw.githubusercontent.com/ktmeaton/plague-phylogeography/master/narratives/images/neolithic-map-1.png" width="80%">
        <p style='text-align: center;'>The Spread of Neolithic Plague [8]</p>
</a>


<!----------------------------------------------------------------------------->
<!--                     4 : LS : Evolution Time Vortex                      -->
<!----------------------------------------------------------------------------->
# [Time Vortex of Evolution](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=tree&l=clock&m=time)
#### To critique the previous maps on the spread of the plague, it is important to consider several statistical problems.

#### To model the past using genetic data, we make and test theories about how evolution proceeds over time.

#### Unfortunately, the plague does not follow the rules of a <b>"molecular clock"</b>, where evolution should occur at a constant rate. Under this model, a 'younger' strain (ex. collected in 2000) should have more mutations than an 'older' strain (ex. collected in 1950).

#### Instead, there are dramatic fluctuations where the bacterium may ["speed up" or "slow down" it's pace of evolution](https://doi.org/10.1073/pnas.1205750110)[<sup>9</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8).

# Regression

#### The visual to the right compares time on the X-axis (the date) with the mutations that have occurred on the Y-axis (divergence). The dots represent plague samples, and ideally they should fall on the dark black line.

#### The bubbles above the black line have <b>more</b> mutations than expected, and those below the black line have <b>fewer</b> mutations than expected.

#### How can we a reconstruct geographic spread over time, with evidence that behaves so chaotically?


<!----------------------------------------------------------------------------->
<!--                     5 : LS : Ecology Picky                              -->
<!----------------------------------------------------------------------------->
# [Plague's Not Picky](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=tree&c=host&legend=open)
#### The solution could be in understanding why this disease does not evolve in a "clock-like" manner. And a clue may lie in considering the ecology of plague.

#### Although plague is primarily a disease of rodents, [virtually all mammals](http://reviverestore.org/wp-content/uploads/2015/02/Gage-and-Kosoy_USGS-Blk-footed-ferret-symp_2006-copy.pdf)[<sup>10</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8) are capable of becoming infected when exposed . The movement of plague between novel hosts and environments may be a key factor in explaining why the rate of evolution changes so suddenly.

#### The accompanying visual is a [phylogenetic tree](https://nextstrain.org/help/general/how-to-read-a-tree), where the bubbles represent plague samples and the connecting lines show their degree of relatedness. The color indicates what mammalian host or flea vector the bacteria was isolated from.

#### No obvious patterns emerge as the colors appear 'randomly' distributed. But perhaps the [rodent subfamily Arvicolinae](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?c=host&d=tree&f_host=Microtus,Neodon,Lasiopodomys) is tentatively associated with extra-long branch lengths (ie. rapid evolution).

#### New perspectives on exploratory data analysis that are ecologically-grounded have great potential to yield greater insight.


<!----------------------------------------------------------------------------->
<!--                     6 : LS : Ecology Human                              -->
<!----------------------------------------------------------------------------->
# [Human Spillover](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?c=host&d=tree,map&f_host=Homo&legend=closed)
#### This ecological fluidity to adapt to different hosts has had devastating consequences for human populations.

#### ['Spillover' events](https://doi.org/10.1038/nrmicro.2017.45)[<sup>11</sup>](https://nextstrain.org/community/narratives/ktmeaton/plague-phylogeography/plagueSCDS2020Remote?n=8), where plague crosses over environmental and species boundaries, has led to human outbreaks all across the globe.

#### There is not just a single strain of plague responsible for human infections. Instead, plague strains from many different lineages throughout the evolutionary tree have been linked to epidemics.

#### However, certain lineages of plague seem to be more 'successful' than others, leading to global pandemics and extensive mortality.

#### The question of why these particular bacteria vary in their virulence potential continues to be one of paramount importance.


<!----------------------------------------------------------------------------->
<!--                     7 : LS : Digital Scholarship                        -->
<!----------------------------------------------------------------------------->
# [Digital Scholarship](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=map)
#### The data in this exhibit all derive from publicly accessible projects available through the [National Centre for Biotechnology Information](https://www.ncbi.nlm.nih.gov/).

#### This exhibit only captures a very small fraction (151 samples) of the available plague datasets that could be harnessed for analysis (1500+ and growing).

#### Method documentation for the project can be found at [Read the Docs](https://plague-phylogeography.readthedocs.io/en/latest/).

#### The underlying code is open-access via a [GitHub Repository](https://github.com/ktmeaton/plague-phylogeography) and is continually developed for reproducibility.

#### Future work will focus on expanding scope, both geographically and temporally, and integrating new theoretical perspectives to shed further light on the mysteries of <b>where</b>, <b>when</b>, and <b>how</b> plague spreads across the globe.

<!----------------------------------------------------------------------------->
<!--                     8 : LS : References Numbered                        -->
<!----------------------------------------------------------------------------->
# [References](https://nextstrain.org/community/ktmeaton/plague-phylogeography/plague150Remote?d=tree,map&legend=closed)
1. Dols, M. W. (1974). [Plague in early Islamic history](https://doi.org/10.2307/600071). *Journal of the American Oriental Society*, 94(3), 371–383.
2. Benedictow, O. J. (2004). [The Black Death, 1346-1353: The Complete History](https://boydellandbrewer.com/the-black-death-1346-1353-the-complete-history.html). Woodbridge, Suffolk, United Kingdom: Boydell Press.
3. Christakos, G., Olea, R. A., Serre, M. L., Yu, H.-L., & Wang, L.-L. (2005). [Interdisciplinary Public Health Reasoning and Epidemic Modelling: The Case of Black Death](https://doi.org/10.1007/3-540-28165-7). Berlin Heidelberg: Springer-Verlag.
4. Skog, L., & Hauska, H. (2013). [Spatial Modeling of the Black Death in Sweden](https://doi.org/10.1111/j.1467-9671.2012.01369.x). *Transactions in GIS*, 17(4), 589–611.
5. Bossak, B. H., & Welford, M. R. (2016). [Spatio-Temporal Characteristics of the Medieval Black Death](https://www.routledge.com/Spatial-Analysis-in-Health-Geography/Kanaroglou-Delmelle/p/book/9781138546615). In P. Kanaroglou & E. D. Delmelle (Eds.), Spatial Analysis in Health Geography (pp. 93–106). Routledge.
6. Laudisoit, A. (2009). [Diversity, ecology and status of potential hosts and vectors of the plague bacillus *Yersinia pestis* (PhD Thesis)](https://doi.org/10.13140/RG.2.2.25362.25281). University of Liège and Antwerpen.
7. Morelli, G., Song, Y., Mazzoni, C. J., Eppinger, M., Roumagnac, P., Wagner, D. M., ... Achtman, M. (2010). [*Yersinia pestis* genome sequencing identifies patterns of global phylogenetic diversity](https://doi.org/10.1038/ng.705). *Nature Genetics*, 42(12), 1140–1143.
8. Rascovan, N., Sjögren, K.-G., Kristiansen, K., Nielsen, R., Willerslev, E., Desnues, C., & Rasmussen, S. (2019). [Emergence and spread of basal lineages of *Yersinia pestis* during the Neolithic Decline](https://doi.org/10.1016/j.cell.2018.11.005). *Cell*, 176(1), 295-305.e10.
9. Cui, Y., Yu, C., Yan, Y., Li, D., Li, Y., Jombart, T., ... Yang, R. (2013). [Historical variations in mutation rate in an epidemic pathogen, *Yersinia pestis*](https://doi.org/10.1073/pnas.1205750110). *Proceedings of the National Academy of Sciences*, 110(2), 577–582.
10. Gage, K., & Kosoy, M. (2006). [Recent trends in plague ecology](http://reviverestore.org/wp-content/uploads/2015/02/Gage-and-Kosoy_USGS-Blk-footed-ferret-symp_2006-copy.pdf). *USG Survey*, 213–231.
11. Plowright, R. K., Parrish, C. R., McCallum, H., Hudson, P. J., Ko, A. I., Graham, A. L., & Lloyd-Smith, J. O. (2017). [Pathways to zoonotic spillover](https://doi.org/10.1038/nrmicro.2017.45). *Nature Reviews Microbiology*, 15(8), 502–510.5
